### PR TITLE
Add a few type hints to the codebase

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
I added some type hints, not all. We can add other type hints later on in other pull requests. I recommend sqash and merge as the merging method as I'm not fond of rewriting my branch history (a bit dangerous).

### CODE REVIEW

File: plum/function.py [hunk 1]
```diff
@@ -14,8 +14,13 @@
 _promised_convert = None
 """function or None: This will be set to :func:`.parametric.convert`."""
 
+# replace this by from typing import Self in python 3.11
+Self = TypeVar("Self")
```
Comment (wesselb) [inline plum/function.py:18]:
```suggestion
# `typing.Self` is available for Python 3.11 and higher.
try:  # pragma: specific no cover 3.11
    from typing import Self
except ImportError:  # pragma: specific no cover 3.7 3.8 3.9 3.10
    Self = TypeVar("Self", bound="Function")
```

How about this?

File: plum/function.py [hunk 2]
```diff
@@ -14,8 +14,13 @@
...
 
-def _convert(obj, target_type):
+SomeExceptionType = TypeVar("SomeExceptionType", bound=Exception)
+
+
+def _convert(obj: Any, target_type: type):
```
Comment (wesselb) [inline plum/function.py:23]:
```suggestion
def _convert(obj: Any, target_type: type) -> Any:
```

File: plum/function.py [hunk 3]
```diff
@@ -391,7 +404,7 @@ def __call__(self, *args, **kw_args):
 
         return _convert(method(*args, **kw_args), return_type)
 
-    def invoke(self, *types):
+    def invoke(self, *types: type):
```
Comment (wesselb) [inline plum/function.py:407]:
```suggestion
    def invoke(self, *types: type) -> Callable:
```

File: plum/parametric.py [hunk 4]
```diff
@@ -114,7 +116,7 @@ def type_parameter(cls):
             )
 
 
-def _default_le_type_par(p_left, p_right):
+def _default_le_type_par(p_left: Union[type, object], p_right: Union[type, object]):
```
Comment (wesselb) [inline plum/parametric.py:119]:
```suggestion
def _default_le_type_par(p_left: Union[type, object], p_right: Union[type, object]) -> bool:
```

File: plum/resolver.py [hunk 5]
```diff
@@ -95,7 +98,7 @@ def doc(self, exclude=None):
         # a newline.
         return "\n\n".join(unique_docs)
 
-    def register(self, signature):
+    def register(self, signature: Signature):
```
Comment (wesselb) [inline plum/resolver.py:101]:
```suggestion
    def register(self, signature: Signature) -> None:
```

**Review by wesselb — COMMENTED** [review]
This is incredibly helpful, @gabrieldemarmiesse! Thank you. :) I've added a few suggestions to add return types wherever the arguments have types.

Let me know when you think this is ready to merge.

**Comment by gabrieldemarmiesse** [comment]
@wesselb you can re-review, I had the time to also change some stuff after running mypy on a bit of it.

type hints cannot be typed at "type", we have to use "object" it's the only public type hint available for it :(

File: plum/function.py (hunk 1 above)
Comment (gabrieldemarmiesse) [reply plum/function.py]: Yes, done

File: plum/resolver.py (hunk 5 above)
Comment (gabrieldemarmiesse) [reply plum/resolver.py]: done

File: plum/function.py (hunk 2 above)
Comment (gabrieldemarmiesse) [reply plum/function.py]: Done

File: plum/parametric.py (hunk 4 above)
Comment (gabrieldemarmiesse) [reply plum/parametric.py]: Done

File: plum/function.py (hunk 3 above)
Comment (gabrieldemarmiesse) [reply plum/function.py]: done

File: plum/dispatcher.py [hunk 6]
```diff
@@ -15,10 +17,10 @@ class Dispatcher:
     """
 
     def __init__(self):
-        self.functions = {}
-        self.classes = {}
+        self.functions: Dict[str, Function] = {}
+        self.classes: Dict[str, Dict[str, Function]] = {}
 
-    def __call__(self, method=None, precedence=0):
+    def __call__(self, method: Optional[Callable] = None, precedence: int = 0):
```
Comment (wesselb) [inline plum/dispatcher.py:23]:
```suggestion
    def __call__(self, method: Optional[Callable] = None, precedence: int = 0) -> Callable:
```

File: plum/function.py [hunk 7]
```diff
@@ -80,22 +88,22 @@ class Function(metaclass=_FunctionMeta):
 
     _instances = []
 
-    def __init__(self, f, owner=None):
+    def __init__(self, f: Callable, owner: Optional[str] = None):
```
Comment (wesselb) [inline plum/function.py:91]:
```suggestion
    def __init__(self, f: Callable, owner: Optional[str] = None) -> None:
```

File: plum/function.py [hunk 8]
```diff
@@ -160,17 +168,19 @@ def __doc__(self):
         return doc if doc else None
 
     @__doc__.setter
-    def __doc__(self, value):
+    def __doc__(self, value: str):
```
Comment (wesselb) [inline plum/function.py:171]:
```suggestion
    def __doc__(self, value: str) -> None:
```

**Comment by wesselb** [comment]
@gabrieldemarmiesse ah, yes, I think you're right that the best we can do is use `object` for type hints. :( That's unfortunate. I think your solution with `TypeHint` is great!!

How would you feel about renaming `TypeHint` to `TypeHintType`, to avoid confusing with `beartype`'s `TypeHint`?

I've added a few more suggestions for return types. (`mypy` only checks functions with returns types, which is why I'm checking them all. :) )

**Comment by gabrieldemarmiesse** [comment]
I'm unsure. TypeHint (the type) will be used a lot more in the codebase than beartype's TypeHint. And TypeHintType has Type in it twice, making it quite hard to understand at first glance and can hinder readability. I would suggest to take one of those three paths:
1) leave the PR untouched and merge it
2) Rename the import `from beartype.door import TypeHint as BeartypeTypeHint`
3) Find a new name for TypeHintType that is very easy to understand and read

**Review by wesselb — APPROVED** [review]
Alright, that's a fair point. I think the current solution is quite reasonable, so let's go with that then!

**Comment by gabrieldemarmiesse** [comment]
Thanks, I'll make more PRs for type hints :)

**Comment by wesselb** [comment]
Thank you, @gabrieldemarmiesse, that's much appreciated. :)

**Pull request merged**